File: assembly-parsing/3d-dna-contig-to-scaffold-tsv.pl
#!/usr/bin/perl

use strict;
use warnings;
use FileHandle;
use Getopt::Long;
use Pod::Usage;

MAIN_CODE: {
	if(scalar @ARGV == 0) { pod2usage(); }

    my $opt = {'assembly' => undef,
			   'label' => undef,
			   'skip_gap' => undef};

	$opt->{'label'} = "HiC_scaffold";

    GetOptions($opt,
	           'assembly=s',
	           'label=s',
	           'skip_gap+') ||  pod2usage();

    my $rh_contigs = parse_contigs($opt->{'assembly'});
    my $fh_IN = new FileHandle;
    $fh_IN->open($opt->{'assembly'}) || die "ERROR: Cannot open assembly file for reading\n";
    my $scaffold_counter = 0;
    while(<$fh_IN>) {
		chomp $_;
		my $i = $_;
		if ($i !~ /^\>/) {
			$scaffold_counter++;
			my $scaffold_name = $opt->{'label'} . '_' . $scaffold_counter;
			my @i = split(/ /, $i);
			foreach my $j (@i) {
				my $strand = '+';
				if ($j < 0) {
					$strand = '-';
				}
				$j = abs($j);
				my $contig_name   = $rh_contigs->{$j}{'contig'};
				my $contig_length = $rh_contigs->{$j}{'length'};
				my @print_me = ($scaffold_name, $contig_name, $strand, $contig_length);
				if ($opt->{'skip_gap'} && $contig_name =~ /^hic_gap_/) {
					next;
				}
				print join("\t", @print_me), "\n";
			}
		}
	}
	$fh_IN->close();
}

sub parse_contigs {
	my $assembly = shift;
	my $fh_IN = new FileHandle;
	$fh_IN->open($assembly) || die "ERROR: Cannot open assembly file for reading\n";
	my %return_me;
	while(<$fh_IN>) {
		chomp $_;
		my $i = $_;
		if ($i =~ /^\>/) {
			$i =~ s/^\>//;
			my @i = split(/ /, $i);
			$return_me{$i[1]}{'contig'} = $i[0];
			$return_me{$i[1]}{'length'} = $i[2];
		}
	}
	$fh_IN->close();
	return \%return_me;
}

=head1 SYNOPSIS

3d-dna-contig-to-scaffold-tsv.pl -assembly *OPTIONAL* -label -skip_gap

File: assembly-parsing/3d-dna-contig-to-scaffold-tsv.py
#!/usr/bin/env python3

# 3d-dna-contig-to-scaffold-tsv.py

import argparse
import csv

parser = argparse.ArgumentParser(description='Open an assembly file and map contigs to HiC Scaffolds')
parser.add_argument('-i', '--infile', help='Assembly file to map', required=True,dest='infile')
args = parser.parse_args()

def parse_contigs():
    scaffold_list = []
    contig_dict = {}
    contig_length = {}
    with open(args.infile, 'r') as infile:
        for lines in infile:
            line = lines.rstrip('\n')
            if line.startswith('>'):
                contig_array = line.strip('>').split(' ')
                name = contig_array[0]
                number = int(contig_array[1])
                length = int(contig_array[2])
                contig_dict[number] = name
                contig_length[number] = length
            else:
                scaffolds = line.split(' ')
                scaffolds = list(map(int, scaffolds)) #turn list into list of integers from list of strings
                scaffold_list.append(scaffolds)
    return contig_dict, contig_length, scaffold_list

def align_scaffolds(contig_dict, contig_length, list):
    scaffold_number = 0
    for l in list:
        scaffold_number += 1
        scaffold_name = "%s_%s"%('HiC_scaffold', scaffold_number)
        for c in l:
            contig_name = contig_dict[abs(c)]
            strand = '+'
            if c < 0:
                strand = '-'
            print("%s\t%s\t%s\t%s"%(scaffold_name, contig_name, strand, contig_length[abs(c)]))
    

def main():
    contig_dict, contig_length, scaffold_list = parse_contigs()
    align_scaffolds(contig_dict, contig_length, scaffold_list)

if __name__ == '__main__':
    main()
